Made Sabertooth Thruster safer; safety fixes

irwineffect opened the pull request.

Commit:
Made Sabertooth Thruster safer; safety fixes

skallaher commented on src/movement/thruster_sabertooth.cpp:
```diff
@@ -22,6 +22,49 @@ void checksum (uint8_t serial_data[4])
   serial_data[3] = (serial_data[0]+serial_data[1]+serial_data[2]) & 127;
 }
 
+void createThrusterPacket (uint8_t serial_data[4], double value, int i)
```
A nice thing here would be to have a comment explaining the structure of the packet so that it is here and the data sheet doesn't need to be referenced when changes are needed. Also, `i` isn't a very descriptive variable name.

bkallaher replied:
A better name for i would be thruster_number

lucy9tran replied:
I do not think i is the thruster number. I will just change it to "index."

bkallaher replied:
It is the thruster number in the mThruster_info data structure.

skallaher commented on src/movement/thruster_sabertooth.cpp:
```diff
@@ -22,6 +22,49 @@ void checksum (uint8_t serial_data[4])
+    }
+    else
+    {
+      if(value < 0)
```
What is the difference between the two cases of `mThruster_info[i].port ==0` and the else? Your commments seem to lead me to believe that there is no difference.

bkallaher replied:
The comments in the else should say port 2

skallaher commented on src/movement/thruster_sabertooth.cpp:
```diff
@@ -84,12 +89,39 @@ void callBack(const robosub::thruster::ConstPtr& msg)
+  uint8_t serial_data[4];
+
+  //loop through 6 thrusters
+  for (int i = 0; i < 6; i++)
```
Why six thrusters? I feel like this could be made more extensible so that if in the future the sabertooth thrusters are used again, but a different number of them. @irwineffect, what do you think?

irwineffect replied:
Yeah, this should be based on the actual number of thrusters.

skallaher commented on src/movement/thruster_sabertooth.cpp:
```diff
@@ -84,12 +89,39 @@ void callBack(const robosub::thruster::ConstPtr& msg)
+    //send package to thrusters
+    mSerial.Write(serial_data, 4);
+
+  //Publish thruster info with name
```
This is not what this does. The comment is inaccurate. Also, it may be more useful to output when each thruster (and which one) is having its timeout set.

bkallaher replied:
This comment seems like an artifact from moving code around

skallaher commented on src/movement/thruster_sabertooth.cpp:
```diff
@@ -84,12 +89,39 @@ void callBack(const robosub::thruster::ConstPtr& msg)
 
 }
 
+void setTimeOut (uint8_t ms_100)
```
One possible thought to avoid confusion is to make this function accept a float or double in seconds, then convert it to the necessary value in uint8_t.

irwineffect replied:
It's set to a uint8_t because the hardware can't have a value greater than 256, this will allow a compile-time check. A comment describing the input is probably a good idea.

skallaher commented on src/movement/thruster_sabertooth.cpp:
```diff
@@ -34,46 +77,8 @@ void callBack(const robosub::thruster::ConstPtr& msg)
 
-      //checksum
-      checksum(serial_data);
+      createThrusterPacket(serial_data, message.data[i], i);
```
`serial_data` has not been declared in this scope. I'm curious as to how Travis missed this.

irwineffect replied:
It is actually declared. It just shows up in the collapsed section

skallaher commented on src/movement/thruster_sabertooth.cpp:
```diff
@@ -117,5 +149,12 @@ int main(int argc, char **argv)
 
   ros::spin();
 
+  uint8_t serial_data[4];
```
If this isn't used outside of the `createThrusterPacket`, some copying could be saved by either passing a pointer or creating the array within the function.

bkallaher replied:
This code shouldn't be here (152-157). All of the work should be (and is) done in the callback function.

irwineffect replied:
@bkallaher, this code here sends the stop signal to all the thrusters when the node shuts down.

skallaher replied:
Did a quick google search and found realized that passing the array this way isn't actually copying the array as it is naturally passed as a pointer. Disregard my previous suggestion to save copying.

bkallaher commented on src/movement/thruster_sabertooth.cpp:
```diff
@@ -84,12 +89,39 @@ void callBack(const robosub::thruster::ConstPtr& msg)
+  }
+
+    //send package to thrusters
+    mSerial.Write(serial_data, 4);
```
This should be inside the for loop.

irwineffect replied:
line 110

lucy9tran replied:
Should it be or should it not be in the loop, James?

bkallaher replied:
With it's current location the code will only ever send to the last thruster as serial data gets overwritten every time through the loop

lucy9tran replied:
Thanks

irwineffect replied:
yep yep

Commit:
Updated based off of comments

Tried to make a general case for different numbers of thrusters
More comments added

Commit:
Update

Package sent to all thrusters
>setTimeOut function

irwineffect committed:
small tweaks for sabertooth thrusters
